### PR TITLE
implemented role rules/validation

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* /bcgov/entity#5312

*Description of changes:*

(This is a port of the changes I made in Edit UI.)

- changed roles to a "group"
- added validation rules for roles
- always enable add/edit Person/Org Done button
- validate form when Done is clicked
- updated Cancel button style (same as Correction design)
- added transparent role icon for proper spacing
- updated/added unit tests
- misc cleanup
- updated version to 1.3.1

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 1]
```diff
@@ -22,8 +22,9 @@
                 ref="addPersonOrgForm"
                 class="appoint-form"
                 v-model="addPersonOrgFormValid"
-                v-on:submit.prevent="addPerson"
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:25]: Same fix done in Edit UI.

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <template>
-  <div>
+  <div id="addEditPersonContainer">
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:2]: The component div should have an ID. The transition should not.

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 3]
```diff
@@ -62,37 +63,46 @@
...
                 <v-row>
                   <v-col cols="4" v-if="isPerson">
-                    <div :class="{'highlightedRole': isRoleLocked(Roles.COMPLETING_PARTY)}">
+                    <div :class="{ 'highlightedRole': isRoleLocked(Roles.COMPLETING_PARTY) }">
                       <v-checkbox
-                        v-model="isCompletingParty"
-                        id='cp-checkbox'
-                        label="Completing Party"
+                        id="cp-checkbox"
+                        v-model="selectedRoles"
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:74]: All 3 checkboxes have the same v-model property (an array), and different "values".

Comment (severinbeauvais) [reply src/components/AddPeopleAndRoles/OrgPerson.vue:74]: This allows the role rules to be simplified (and also, the rules didn't update properly with 3 different v-model properties).

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 4]
```diff
@@ -62,37 +63,46 @@
...
+                    <div :class="{ 'highlightedRole': isRoleLocked(Roles.INCORPORATOR) || isOrg }"
                     >
                       <v-checkbox
-                        v-model="isIncorporator"
+                        v-model="selectedRoles"
+                        :value="Roles.INCORPORATOR"
                         :label="incorporatorLabel"
-                        :disabled="isRoleLocked(Roles.INCORPORATOR) ||
-                        orgPerson.officer.partyType === IncorporatorTypes.CORPORATION"
+                        :disabled="isRoleLocked(Roles.INCORPORATOR) || isOrg"
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:90]: "isOrg" already did this check (and used null chaining for safety), so this cleans up the code.

File: src/components/AddPeopleAndRoles/PeopleAndRoles.vue [hunk 5]
```diff
@@ -8,17 +8,17 @@
     <ul>
       <li>
         <v-icon v-if="hasRole(Roles.COMPLETING_PARTY, 1, 'EXACT')" color="blue" class="cp-valid">mdi-check</v-icon>
-        <v-icon v-else-if="showErrors" color="red" class="cp-invalid">mdi-close</v-icon>
+        <v-icon v-else :color="showErrors ? 'red': 'transparent'" class="cp-invalid">mdi-close</v-icon>
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/PeopleAndRoles.vue:11]: Sample output without errors:

image

Comment (severinbeauvais) [reply src/components/AddPeopleAndRoles/PeopleAndRoles.vue:11]: Sample ouput with errors shown:

image

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 6]
```diff
@@ -380,10 +413,6 @@ export default class OrgPerson extends Mixins(EntityFilterMixin, CommonMixin) {
     return (this.orgPerson.roles.some(party => party.roleType === role) && this.activeIndex === -1)
   }
 
-  private removePerson (): void {
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:383]: I simply moved "emitRemovePersonEvent(activeIndex)" into the template.

File: src/components/AddPeopleAndRoles/OrgPerson.vue [hunk 7]
```diff
@@ -287,24 +320,24 @@ export default class OrgPerson extends Mixins(EntityFilterMixin, CommonMixin) {
...
         this.emitReassignCompletingPartyEvent()
       }
       const person: OrgPersonIF = this.addPerson()
       this.emitPersonInfo(person)
-      this.resetAddPersonData(false)
+      this.resetAddPersonData(false) // don't emit event
     }
   }
 
-  private isFormValid (): boolean {
```
Comment (severinbeauvais) [inline src/components/AddPeopleAndRoles/OrgPerson.vue:300]: I just moved this down to be with the other getters.

File: tests/unit/OrgPerson.spec.ts [hunk 8]
```diff
@@ -164,18 +187,18 @@ describe('Org Person component', () => {
   it('Loads the component and sets data for person', async () => {
     const wrapper: Wrapper<OrgPerson> = createComponent(validPersonData, -1, 0, null)
     expect(wrapper.vm.$data.orgPerson).toStrictEqual(validPersonData)
-    expect(wrapper.vm.$data.isIncorporator).toBe(false)
-    expect(wrapper.vm.$data.isDirector).toBe(true)
-    expect(wrapper.vm.$data.isCompletingParty).toBe(true)
+    expect((wrapper.vm as any).isIncorporator).toBe(false)
```
Comment (severinbeauvais) [inline tests/unit/OrgPerson.spec.ts:190]: "isXXX" are now getters (methods) instead of properties (data).

**Review by lekshmimallika-aot — APPROVED** [review]

**Review by cameron-freshworks — APPROVED** [review]

**Pull request merged**